src/sources/opentelemetry/http.rs

```diff
@@ -0,0 +1,155 @@
+use std::net::SocketAddr;
+
+use bytes::Bytes;
+use futures_util::FutureExt;
+use http::StatusCode;
+use prost::Message;
+use snafu::Snafu;
+use tracing::Span;
+use vector_core::{
+    event::{BatchNotifier, BatchStatus},
+    ByteSizeOf,
+};
+use warp::{filters::BoxedFilter, reject::Rejection, reply::Response, Filter, Reply};
+
+use crate::{
+    event::Event,
+    internal_events::{EventsReceived, HttpBytesReceived, StreamClosedError},
+    opentelemetry::LogService::{ExportLogsServiceRequest, ExportLogsServiceResponse},
+    shutdown::ShutdownSignal,
+    sources::util::{decode, ErrorMessage},
+    tls::MaybeTlsSettings,
+    SourceSender,
+};
+
+use super::{reply::protobuf, status::Status};
+
+#[derive(Clone, Copy, Debug, Snafu)]
+pub(crate) enum ApiError {
+    BadRequest,
+    ServerShutdown,
+}
+
+impl warp::reject::Reject for ApiError {}
+
+pub(crate) async fn run_http_server(
+    address: SocketAddr,
+    tls_settings: MaybeTlsSettings,
+    filters: BoxedFilter<(Response,)>,
+    shutdown: ShutdownSignal,
+) -> crate::Result<()> {
+    let span = Span::current();
+    let listener = tls_settings.bind(&address).await?;
+    let routes = filters
+        .with(warp::trace(move |_info| span.clone()))
+        .recover(|r: Rejection| async move {
+            if let Some(e_msg) = r.find::<ErrorMessage>() {
+                let reply = protobuf(Status {
+                    code: 2, // UNKNOWN - OTLP doesn't require use of status.code, but we can't encode a None here
+                    message: e_msg.message().into(),
+                    ..Default::default()
+                });
+
+                Ok(warp::reply::with_status(reply, e_msg.status_code()))
+            } else {
+                // other internal error - will return 500 internal server error
```

Comment:
This isn't correct per spec, but I was struggling return a properly encoded `Status` in the body here and opted for logging the unhandled `Rejection` - which we could then fix.

Reply:
I think I'm missing something. Couldn't you do the same thing you do above with `protobuf(Status{ ... })`?

Reply:
Is this something that still needs to be followed up on? Or documented per a comment ?

Reply:
Playing with some suggestions from @jszwedko (without luck so far), I'll timebox this to not interrupt other work. It can also be improved as I add the other datatypes, it should function correctly - just not be "per spec" in terms of the Response Body.

Reply:
Thanks to @jszwedko's guidance, this is resolved!